### PR TITLE
ICU-22825 Fix memLeak during error in tznames_impl.cpp

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22825
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Review by richgillam — APPROVED** [review]

File: icu4c/source/i18n/tznames_impl.cpp [hunk 1]
```diff
@@ -731,6 +731,7 @@ class ZNames : public UMemory {
                 nameinfo->type = getTZNameType(static_cast<UTimeZoneNameTypeIndex>(i));
                 trie.put(name, nameinfo, status); // trie.put() takes ownership of the key
                 if (U_FAILURE(status)) {
+                    uprv_free(nameinfo);
```
Comment (markusicu) [inline icu4c/source/i18n/tznames_impl.cpp:734]: If trie.put() truly takes ownership, then it should do so even when something goes wrong. In that case, we would now get a double-delete problem.
Please check.
@yumaoka @aheninger FYI

**Review by markusicu — CHANGES_REQUESTED** [review]

File: icu4c/source/i18n/tznames_impl.cpp (hunk 1 above)
Comment (FrankYFTang) [reply icu4c/source/i18n/tznames_impl.cpp:734]: you are right. I think the real issue is the TextTrieMap::put() implement called the deleter on the wrong variable. it should call the deleter to delete the value, not the key

**Review by richgillam — APPROVED** [review]
I _think_ this is right.  I was concerned that you might need to take ownership of both the key and the value, but the documentation says this particular function is for cases where the key is in static memory owned by somebody else (the resource bundle), so I think you're doing the right thing.

**Comment by FrankYFTang** [comment]
@sffc could you double check since you are the one who wrote the origional code.

**Review by sffc — APPROVED** [review]
The minimal fix would be to change `fValueDeleter((void*) key);` to `fValueDeleter(value);`

The rest of the refactor looks fine.

**Comment by FrankYFTang** [comment]
> The minimal fix would be to change `fValueDeleter((void*) key);` to `fValueDeleter(value);`

no, that will not be enugh because for the last line of 
fLazyContents->addElement(value, status); in the origional line 255
we also need a "if fail if deleter delete block" in case the return status faile also. That is the leak the reporter detected.

**Pull request merged**